swiftPackages.*: 5.8 -> 5.10.1

Samasaur1 opened the pull request:
Updates Swift and the corresponding tools from 5.8 to 5.10.1, which also means updating the propagated SDK from `apple-sdk_13` to `apple-sdk_14`. This in turn allows building Swift packages that rely on APIs introduced in macOS 14 Sonoma.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Samasaur1 committed:
swiftPackages.swiftpm: use newer swift-crypto to fix missing symbols

This is 3.0.0 -> 3.15.1, so it's even semver compatible

Samasaur1 committed:
swiftPackages.*: bump SDK to 14

Samasaur1 committed:
swiftPackages.swift: fix patches that no longer apply cleanly

Samasaur1 committed:
swiftPackages.{sourcekit-lsp,swift-format}: link libdispatch.so

In my experience this solution seems to be necessary on almost all if
not all Swift packages on Linux

Samasaur1 committed:
swiftPackages.*: update sources from 5.8 to 5.10.1

Samasaur1 committed:
swiftPackages.swift: drop upstreamed patches

Samasaur1 committed:
swiftPackages.swift: enable remaining compiler features

There are remaining features that are not enabled, but they depend on a
host Swift with which to compile.

Samasaur1 committed:
swiftPackages.swiftpm: add bootstrap derivations for new dependencies

Samasaur1 committed:
swiftPackages.swiftpm: fix patches that no longer apply cleanly

Samasaur1 committed:
swiftPackages.swiftpm: patch swift-crypto as needed

Samasaur1 committed:
swiftPackages.swift-corelibs-foundation: drop upstreamed patches

Technically the first of these wasn't upstreamed and was instead
superceded by
https://github.com/swiftlang/swift-corelibs-foundation/commit/8d1a6794206637fee67f4a955853b0d1771746e2
but that's close enough.

Samasaur1 committed:
swiftPackages.swift: clang 15 -> 16

Samasaur1 committed:
swiftPackages.*: swap from replace to replace-fail

Samasaur1 committed:
swiftPackages.swift: check for modern CXX interop flag in wrapper

This check is what determines whether the appropriate CFLAGS are set so
that Swift can find the C++ headers it needs for interop

Samasaur1 committed:
swiftPackages.*: update generated SwiftPM dependencies

This includes adding new dependencies when appropriate

Samasaur1 committed:
swiftPackages.swiftpm: drop upstreamed patches

Some of these were not technically upstreamed but all have been fixed

Samasaur1 committed:
swiftPackages.{sourcekit-lsp,swift-format}: drop upstreamed patches

I'm grouping this together because it's the same patch

Samasaur1 committed:
swiftPackages.swift: update cxx-interop-test

Samasaur1 committed:
swiftPackages.swift{,-corelibs-libdispatch}: fix build with modern clang

We apply this patch to the copy of libdispatch that's copied into Swift
and to the standalone build.

Samasaur1 committed:
swiftPackages.swift-driver: fix patches that no longer apply cleanly

Samasaur1 committed:
teams/swift: add samasaur

WeetHet commented:
> I can finally swift..
Afaik this is just a temporary solution until the swift rework done by @reckenrode is done (probably in 26.05). It's basically only done so that the 13.3 SDK can be dropped

Samasaur1 committed:
swiftPackages.swift: fix patchPhase indentation

Samasaur1 committed:
swiftPackages.swift: patch CMake to import missing module

Samasaur1 committed:
swiftPackages.swift: drop SysRoot replacements that fail to apply

These replacements didn't apply cleanly to Swift 5.8 anyway, we just
didn't notice. They don't appear to cause any problems and regardless,
the goal is "no more broken than before"

emilazy reviewed (commented):
Looks good, thank you! One patch‐splitting mishap that I’d like to see fixed and some other non‐blocking things.

emilazy reviewed (approved):
Thanks!
